File: SortVision/public/code/bucket/julia/bucketSort.jl
# Bucket Sort (for floats in [0, 1))
# Time: Avg O(n + k), Worst O(n^2), Space: O(n + k)
function bucket_sort!(arr::Vector{Float64})
    n = length(arr)
    if n <= 1
        return arr
    end

    buckets = [Float64[] for _ in 1:n]

    for x in arr
        idx = min(n, floor(Int, x * n) + 1)
        push!(buckets[idx], x)
    end

    k = 1
    for b in buckets
        sort!(b)
        for x in b
            arr[k] = x
            k += 1
        end
    end
    return arr
end

arr = [0.42, 0.32, 0.23, 0.52, 0.25, 0.47, 0.51]
println(bucket_sort!(arr))

File: SortVision/public/code/bucket/java/bucketSort.java
import java.util.ArrayList;
import java.util.Collections;

public class BucketSort {
    // Bucket Sort (for floats in [0, 1))
    // Time: Avg O(n + k), Worst O(n^2), Space: O(n + k)
    public static void bucketSort(float[] arr) {
        if (arr == null || arr.length <= 1) return;

        int n = arr.length;
        ArrayList<Float>[] buckets = new ArrayList[n];
        for (int i = 0; i < n; i++) buckets[i] = new ArrayList<>();

        for (float x : arr) {
            int idx = Math.min(n - 1, (int)(x * n));
            buckets[idx].add(x);
        }

        int k = 0;
        for (ArrayList<Float> bucket : buckets) {
            Collections.sort(bucket);
            for (float x : bucket) arr[k++] = x;
        }
    }

    public static void main(String[] args) {
        float[] arr = {0.42f, 0.32f, 0.23f, 0.52f, 0.25f, 0.47f, 0.51f};
        bucketSort(arr);
        for (float x : arr) System.out.print(x + " ");
        System.out.println();
    }
}
